### PR TITLE
Optimize nn.Module __call__ fast path for dynamo

### PR DESCRIPTION
**Pull request by wconstab**
This PR optimizes the guards overhead introduced by dynamo tracing module forward hooks.

It can and maybe should be followed by a wider change proposed by @voznesenskym to optimize specialized nnmodules by 'observing' any user mutations and directly invalidating the root guard, obviating the need to install other nnmodule guards.  (But this observer change seems more involved...)

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #95971
* __->__ #95931

Idea: maintain a flag, and keep it up to date whenever adding or
removing hooks. Use the flag rather than dict checks to enter the call fast path. 
  - need to extend RemovableHandle to keep a ref to nnModule so it can update the flag on removal.
  - also need to handle the flag in ScriptModule which still uses the python call impl when called from python.

cc @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

**Review by ezyang — APPROVED** [review]
Assuming the perf is good, this LGTM

File: torch/nn/modules/module.py [hunk 1]
```diff
@@ -1187,10 +1209,11 @@ def register_full_backward_pre_hook(
                 ``handle.remove()``
 
         """
-        handle = hooks.RemovableHandle(self._backward_pre_hooks)
+        handle = hooks.RemovableHandle(self._backward_pre_hooks, module=self)
```
[the remark this answers is not in the document]
Comment (wconstab) [reply torch/nn/modules/module.py:1212]: it's for calling module._update_has_hooks when RemovableHandle removes a hook from a module

**Review by voznesenskym — APPROVED** [review]

**Comment by wconstab** [comment]
perf looks good:
https://docs.google.com/spreadsheets/d/1k4l4Po-c0dycheaQ0E0aWn3woTsvCL1w98Vij6f-8VQ/edit#gid=354152790

i see about the same 2-3% geomean speedup that i saw as a regression initially with hooks support

**Comment by vadimkantorov** [comment]
My proposal is to also allow the user set/express no-hooks-will-be-added flag on a module. This should enable more aggressive inplace optimization.

**Comment by wconstab** [comment]
> My proposal is to also allow the user set/express no-hooks-will-be-added flag on a module. This should enable more aggressive inplace optimization.

can you say more about what optimization this 'no-hooks' flag would enable?

I believe if hooks are not used, the only impact their infra has is that dynamo adds guards that ensure no hooks are present.  This should still allow any optimizations downstream to assume there will be no hooks.

My PR basically reduces the number of guards for the no-hook case, but more importantly, it avoids any 'type check' guards which are slower to run, and only uses 'bool' constant guards which are fast.

**Comment by vadimkantorov** [comment]
@wconstab There's a long discussion in https://github.com/pytorch/pytorch/issues/23756 and in https://github.com/pytorch/pytorch/issues/23756#issuecomment-792338938

Basically, if we promise that no hooks will be called, PyTorch can safely fuse and do inplace optimizations, especially with invertible modules

File: torch/nn/modules/module.py [hunk 2]
```diff
@@ -94,6 +94,16 @@ def __setstate__(self, state: Dict):
 _global_is_full_backward_hook: Optional[bool] = None
 _global_forward_pre_hooks: Dict[int, Callable] = OrderedDict()
 _global_forward_hooks: Dict[int, Callable] = OrderedDict()
+_has_global_hooks: bool = False
+
+def _update_has_global_hooks():
```
Comment (albanD) [inline torch/nn/modules/module.py:99]: Shouldn't this also be called when handle.remove() is called?
If you don't care about accurate tracking, you can just set it to True and that's it?

File: torch/nn/modules/module.py [hunk 3]
```diff
@@ -1494,10 +1526,10 @@ def _slow_forward(self, *input, **kwargs):
     def _call_impl(self, *args, **kwargs):
         forward_call = (self._slow_forward if torch._C._get_tracing_state() else self.forward)
         # If we don't have any hooks, we want to skip the rest of the logic in
-        # this function, and just call forward.
-        if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks
-                or _global_backward_pre_hooks or _global_backward_hooks
-                or _global_forward_hooks or _global_forward_pre_hooks):
+        # this function, and just call forward.  It's slow for dynamo to guard on the state
```
Comment (albanD) [inline torch/nn/modules/module.py:1529]: Do you have benchmark for that? I'm really curious how checking two bools is significantly faster than checking if dicts are empty?
Do you really care about 10ns here??
```
In [1]: a, b, c, d = {}, {}, {}, {}

In [2]: e, f = False, False

In [3]: %timeit a or b or c or d
22.9 ns ± 0.202 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [4]: %timeit e or f
11.1 ns ± 0.0778 ns per loop (mean ± std. dev. of 7 runs, 100,000,000 loops each)
```

File: torch/nn/modules/module.py [hunk 4]
```diff
@@ -1494,10 +1526,10 @@ def _slow_forward(self, *input, **kwargs):
...
+        # of all these hook dicts individually, so instead it can guard on 2 bools and we just
+        # have to promise to keep them up to date when hooks are added or removed via official means.
+        if not self._has_hooks and not _has_global_hooks:
```
Comment (albanD) [inline torch/nn/modules/module.py:1532]: It's not ok to assume that `_has_hooks` is available as some people pickle their Module directly. You need to update the `__setstate__`  function to properly populate this field if it is not in the serialized state.

File: torch/utils/hooks.py [hunk 5]
```diff
@@ -14,14 +14,21 @@ class RemovableHandle:
...
     id: int
     next_id: int = 0
 
-    def __init__(self, hooks_dict: Any, *, extra_dict: Any = None) -> None:
+    def __init__(self, hooks_dict: Any, *, extra_dict: Any = None, module=None) -> None:
         self.hooks_dict_ref = weakref.ref(hooks_dict)
         self.id = RemovableHandle.next_id
+
+        # TODO: we don't pickle/unpickle this field, which means the 'update_has_hooks'
+        # functionality (which is an optimization) decays after pickling.  Can we fix this?
```
Comment (albanD) [inline torch/utils/hooks.py:29]: It is NOT an optimization, you rely on it to be correct to have the right behavior! That sounds bad?

File: torch/utils/hooks.py [hunk 6]
```diff
@@ -61,6 +74,8 @@ def __setstate__(self, state) -> None:
             if len(state) < 3
             else weakref.ref(OrderedDict() if state[2] is None else state[2])
         )
+        # TODO can we actually restore module_ref after unpickling? Do we care?
```
Comment (albanD) [inline torch/utils/hooks.py:77]: Please do not merge PRs with random TODOs in utils or nn. This code is exercised in many ways compared to dynamo code and this kind of things really matter.

File: torch/nn/modules/module.py (hunk 3 above)
Comment (wconstab) [reply torch/nn/modules/module.py:1529]: yes, i ran benchmarks.  the issue really came to dynamo having to run additional check_type_id() on each dict in addition to confirming each dict is empty.

the dynamo guard for len(dict) is somewhat slow, but the dynamo guard for bool(dict) is fast.

File: torch/utils/hooks.py (hunk 5 above)
Comment (wconstab) [reply torch/utils/hooks.py:29]: Here's why i thought of it as an optimization (maybe i missed a case?)

1) you save a TS module that _has_ hooks and you pickle a RemovableHandle associated with a hook
2) load the scriptmodule and the handle
3a) run the module: the module correctly recomputes the value of '_has_hooks' on load
3b) remove the hook using the unpickled handle and then run the module:  unclear if remove even works by itself, after save/load? i doubt it.  But regardless, if the module still thinks it '_has_hooks' and the hook is removed, all that happens is the fast path will be skipped and the slow path will check the hooks dicts

File: torch/nn/modules/module.py (hunk 4 above)
Comment (wconstab) [reply torch/nn/modules/module.py:1532]: ok, good point. i believe i should add a test case for pickle/unpickle and ensure that __setstate__ correctly enforces that _update_has_hooks gets installed and run.

would that address your concern?

File: torch/nn/modules/module.py (hunk 2 above)
Comment (wconstab) [reply torch/nn/modules/module.py:99]: it is called by handle.remove()

tagging the code below where it is called.

File: torch/utils/hooks.py [hunk 7]
```diff
@@ -40,6 +47,12 @@ def remove(self) -> None:
             if extra_dict is not None and self.id in extra_dict:
                 del extra_dict[self.id]
 
+        if self.module_ref is not None:
+            module = self.module_ref()
+            if module is not None:
+                module._update_has_hooks()
+        torch.nn.modules.module._update_has_global_hooks()
```
Comment (wconstab) [inline torch/utils/hooks.py:54]: @albanD this is where I call _update_has_global_hooks()

File: torch/nn/modules/module.py (hunk 3 above)
Comment (albanD) [reply torch/nn/modules/module.py:1529]: But don't you have to also call check_type_id for boolean fields? 


> the dynamo guard for len(dict) is somewhat slow, but the dynamo guard for bool(dict) is fast.

So would moving dynamo to properly use the bool(dict) guard solve this?

File: torch/nn/modules/module.py (hunk 4 above)
Comment (albanD) [reply torch/nn/modules/module.py:1532]: fixing setstate will work yes.
Unfortunately testing this is challenging indeed :/

Comment (wconstab) [reply torch/nn/modules/module.py:1532]: Fix via: https://github.com/pytorch/pytorch/pull/96224

File: torch/utils/hooks.py (hunk 6 above)
Comment (albanD) [reply torch/utils/hooks.py:77]: Yes see comments below. There are clear BC issues. I think there is some mis-categorization of this as an "optimization" (implying that it's fine if we're not doing a perfect job) even though it is actually critical for correctness (and so we have to do a perfect job here).

**Comment by gchanan** [comment]
> i see about the same 2-3% geomean speedup that i saw as a regression initially with hooks support

Can you open up your spreadsheet? :).

**Pull request closed without merging**